### ISSUE TITLE
Safari navigate to URL when opening deep link hangs sometimes

### ISSUE THREAD

**Issue by jamesvanhorn** [appium/appium#15575]
## The problem

When navigating to a deep link URL in Safari (in order to open a native app), the call to /url endpoint does not return in some cases, hanging indefinitely.

## Environment

* Appium version (or git revision) that exhibits the issue: 1.21.0
* Last Appium version that did not exhibit the issue (if applicable): N/A
* Desktop OS/version used to run Appium: Mac Big Sur
* Node.js version (unless using Appium.app|exe): v15.5.1
* Npm or Yarn package manager: Npm 6.14.11
* Mobile platform/version under test: iOS 14
* Real device or emulator/simulator: Real
* Appium CLI or Appium.app|exe: CLI

## Details

Launch safari -> navigate to a universal link (https link associated with our app, not the specific deep link with special app identifier. We test those in iOS simulator ) -> App opens as intended 100% of the time, however some of the time, the /url commands never returns (or times out or anything), causing the test to hang forever.

## Link to Appium logs

https://gist.github.com/jamesvanhorn/de6b64799737f47c7916c4accc0b118e

Of note, a log of a 'good' run looks identical, except the /url POST call returns as so....

```
[debug] [RemoteDebugger] Received data response from send (id: 56): '{}'
[debug] [RemoteDebugger] Sending to Web Inspector took 22ms
[debug] [W3C (276c85a2)] Responding to client with driver.setUrl() result: null
[HTTP] <-- POST /wd/hub/session/276c85a2-cfad-4af4-8f4d-c96dbff8a650/url 200 1606 ms - 14
```

The app always opens as expected, so even if the call to get(URL) threw an exception or timed out that would be fine, I could work around it. The main issue is it hangs forever until I manually kill the test. Any additional debugging I can run with for some insights? Assuming it has something to do with Safari moving to the background and our app to the foreground.

## Code To Reproduce Issue [ Good To Have ]
driver.get(universalLinkUrl);

**Comment by jlipps** [comment]
If you want to call driver.get on a native (non-web) link or deep link, then make sure you switch to the native context first. Otherwise the URL navigation is getting sent as a javascript atom, which is probably not going to work.

**Comment by jlipps** [comment]
I.e., driver.get is overloaded for both web and native contexts. driver.get in web contexts is for websites. driver.get in native contexts is for deep links. Best not to try to swap responsibilities.

**Comment by jamesvanhorn** [comment]
Hmmm, I believe I do want to call get in a web context so it loads in Safari and redirects to the correct page in the app (that is the functionality I am testing). If I switch context first, it just opens Siri and does nothing (real device). 

Cant test on simulator as http(s) redirects from Safari to native apps don't work. 

Note this issue is related to https://weather.com/blah links, _NOT_ twcweather://blah links. twcweather:// links work fine in native context on simulators as you suggested (same issue with Siri opening and not doing anything on a real device for twcweather:// links)

**Comment by mykola-mokhnach** [comment]
As a possible workaround you could try to paste the link into Safari's address bar in native context without switching to web context at all.

**Comment by jamesvanhorn** [comment]
Good call, having success so far doing this all from native context. A little slower to manually manipulate Safari to click/type/enter the URL, but better than nothing.

**Comment by jlipps** [comment]
Yeah, I think Safari has some security features that prevent certain actions (like opening new windows or apps) being triggered from JavaScript. You could also hunt for that setting and disable it. But doing the exact thing the user does (putting in the native URL bar as you mentioned) is going to potentially be the most reliable, as well as work on real devices as well as simulators.

**Comment by mykola-mokhnach** [comment]
I believe recent xctest features have improved this behaviour nowadays
